protocols/perf/src/protocol.rs

```diff
@@ -0,0 +1,187 @@
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::time::Instant;
+
+use futures::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+use crate::{RunParams, RunTimers};
+
+const BUF: [u8; 1024] = [0; 1024];
+
+pub async fn send_receive<S: AsyncRead + AsyncWrite + Unpin>(
+    params: RunParams,
+    mut stream: S,
+) -> Result<RunTimers, std::io::Error> {
+    let RunParams {
+        to_send,
+        to_receive,
+    } = params;
+
+    let mut receive_buf = vec![0; 1024];
+
+    stream.write_all(&(to_receive as u64).to_be_bytes()).await?;
+
+    let write_start = Instant::now();
+
+    let mut sent = 0;
+    while sent < to_send {
+        let n = std::cmp::min(to_send - sent, BUF.len());
+        let buf = &BUF[..n];
+
+        sent += stream.write(buf).await?;
+    }
+
+    stream.close().await?;
+
+    let write_done = Instant::now();
+
+    let mut received = 0;
+    while received < to_receive {
+        received += stream.read(&mut receive_buf).await?;
+    }
+
+    let read_done = Instant::now();
+
+    Ok(RunTimers {
+        write_start,
+        write_done,
+        read_done,
+    })
+}
+
+pub async fn receive_send<S: AsyncRead + AsyncWrite + Unpin>(
+    mut stream: S,
+) -> Result<(), std::io::Error> {
+    let to_send = {
+        let mut buf = [0; 8];
+        stream.read_exact(&mut buf).await?;
+
+        u64::from_be_bytes(buf) as usize
+    };
+
+    let mut receive_buf = vec![0; 1024];
+    while stream.read(&mut receive_buf).await? != 0 {}
+
+    let mut sent = 0;
+    while sent < to_send {
+        let n = std::cmp::min(to_send - sent, BUF.len());
+        let buf = &BUF[..n];
+
+        sent += stream.write(buf).await?;
+    }
+
+    stream.close().await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{executor::block_on, AsyncRead, AsyncWrite};
+    use std::{
+        pin::Pin,
+        sync::{Arc, Mutex},
+        task::Poll,
+    };
+
+    #[derive(Clone)]
+    struct DummyStream {
+        inner: Arc<Mutex<DummyStreamInner>>,
+    }
```

Comment:
Something like this, a bit more polished would make a great utility that we could use for all sorts of testing.